scio-extra/src/main/scala/com/spotify/scio/extra/voyager/package.scala

```diff
@@ -0,0 +1,291 @@
+      this.asVoyager(path, distanceMeasure, storageDataType, dim, ef, m)
+    }
+
+    /**
+     * Write the key-value pairs of this SCollection as a Voyager index to a temporary location,
+     * building the index using the parameters specified and then loading the reader into a side
+     * input.
+     *
+     * @param spaceType
+     *   The measurement for computing distance between entities. One of Euclidean, Cosine or Dot
+     *   (inner product).
+     * @param storageType
+     *   The Storage type of the vectors at rest. One of Float8, Float32 or E4M3.
+     * @param dim
+     *   Number of dimensions in vectors.
+     * @param ef
+     *   The size of the dynamic list of neighbors used during construction time. This parameter
+     *   controls query time/accuracy tradeoff. More information can be found in the hnswlib
+     *   documentation https://github.com/nmslib/hnswlib.
+     * @param m
+     *   The number of outgoing connections in the graph.
+     * @return
+     *   A SideInput with a [[VoyagerReader]]
+     */
+    @experimental
+    def asVoyagerSideInput(
+      spaceType: SpaceType,
+      storageType: StorageDataType,
+      dim: Int,
+      ef: Long = 200L,
+      m: Long = 16L
+    ): SideInput[VoyagerReader] =
+      self
+        .asVoyager(spaceType, storageType, dim, ef, m)
+        .asVoyagerSideInput(spaceType, storageType, dim)
+  }
+
+  /**
+   * Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with Voyager methods.
+   */
+  implicit class AnnoySCollection(@transient private val self: SCollection[VoyagerUri])
+      extends AnyVal {
+
+    /**
+     * Load the Voyager index stored at [[VoyagerUri]] in this
+     * [[com.spotify.scio.values.SCollection SCollection]].
+     *
+     * @param spaceType
+     *   The measurement for computing distance between entities. One of Euclidean, Cosine or Dot
+     *   (inner product).
+     * @param storageType
+     *   The Storage type of the vectors at rest. One of Float8, Float32 or E4M3.
+     * @param dim
+     *   Number of dimensions in vectors.
+     * @return
+     *   SideInput[VoyagerReader]
+     */
+    @experimental
+    def asVoyagerSideInput(
+      spaceType: SpaceType,
+      storageType: StorageDataType,
+      dim: Int
+    ): SideInput[VoyagerReader] = {
+      val view = self.applyInternal(View.asSingleton())
+      new VoyagerSideInput(view, spaceType, storageType, dim)
+    }
+  }
+
+  /**
+   * Construction for a VoyagerSide input that leverages a synchronized map to ensure that the
+   * reader is only loaded once per [[VoyagerUri]].
+   */
+  private class VoyagerSideInput(
+    val view: PCollectionView[VoyagerUri],
+    distanceMeasure: SpaceType,
+    storageType: StorageDataType,
+    dim: Int
+  ) extends SideInput[VoyagerReader] {
+    override def get[I, O](context: DoFn[I, O]#ProcessContext): VoyagerReader = {
+      val uri = context.sideInput(view)
+      VOYAGER_URI_MAP.synchronized {
```

Comment:
can you explain the synchronized map use case a bit? I guess it would be if a user tried to load two Voyager SideInputs w/ the same URI at different nodes of the job graph... is loading the reader a really expensive operation itself, like does it load everything into memory at construction time, or is it just a pointer to a URI?

Reply:
Yes the loading is really expensive and having multiple nodes / threads try to create a reader concurrently will cause the job to OOM. This sort of alleviates that and lets it be a singleton of uri -> reader at least for constructing and loading it into memory.